src/bit-fields/src/bit.rs

```diff
@@ -0,0 +1,817 @@
+                bool::from(self)
+            }
+
+            // These values are only evaluated at compile-time, thus a failure can only occur at
+            // compile-time and would be immediately obvious. Thus it is safe to use arithmetic
+            // here.
+            #[doc = concat!("
+                Set the bit to 0.
+                ```
+                use bit_fields::BitMut;
+                let mut x = 5", stringify!($x), ";
+                BitMut::<_,0>(&mut x).off();
+                assert_eq!(x,4);
+                BitMut::<_,1>(&mut x).off();
+                assert_eq!(x,4);
+                BitMut::<_,2>(&mut x).off();
+                assert_eq!(x,0);
+                ```
+            ")]
+            #[allow(clippy::integer_arithmetic)]
+            #[inline]
+            pub fn off(&mut self) {
+                *self.0 &= !Self::MASK;
+            }
+
+            #[doc = concat!("
+                Returns if the bit is 0.
+                ```
+                use bit_fields::BitMut;
+                let mut x = 5", stringify!($x), ";
+                assert_eq!(BitMut::<_,0>(&mut x).is_off(),false);
+                assert_eq!(BitMut::<_,1>(&mut x).is_off(),true);
+                assert_eq!(BitMut::<_,2>(&mut x).is_off(),false);
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub fn is_off(&self) -> bool {
+                !bool::from(self)
+            }
+
+            // These values are only evaluated at compile-time, thus a failure can only occur at
+            // compile-time and would be immediately obvious. Thus it is safe to use arithmetic
+            // here.
+            #[doc = concat!("
+                Flips the value of the bit.
+                ```
+                use bit_fields::BitMut;
+                let mut x = 5", stringify!($x), ";
+                BitMut::<_,0>(&mut x).flip();
+                assert_eq!(x,4);
+                BitMut::<_,1>(&mut x).flip();
+                assert_eq!(x,6);
+                BitMut::<_,2>(&mut x).flip();
+                assert_eq!(x,2);
+                ```
+            ")]
+            #[allow(clippy::integer_arithmetic)]
+            #[inline]
+            pub fn flip(&mut self) {
+                *self.0 ^= Self::MASK;
+            }
+
+            #[doc = concat!("
+                Sets the bit.
+                ```
+                use bit_fields::BitMut;
+                let mut x = 5", stringify!($x), ";
+                BitMut::<_,0>(&mut x).set(false);
+                assert_eq!(x,4);
+                BitMut::<_,1>(&mut x).set(true);
+                assert_eq!(x,6);
+                BitMut::<_,2>(&mut x).set(false);
+                assert_eq!(x,2);
+                ```
+            ")]
+            #[inline]
+            pub fn set(&mut self, x: bool) {
+                if x {
+                    self.on()
+                } else {
+                    self.off()
+                }
```

Comment:
You can set bit without branches:
```rust
*self.0 = (*self.0 & !Self::MASK) | (x as $x) << P;
```

Reply:
I would argue that is much less readable. I can see we may want to do this if the performance of this code is important. So the question becomes: is this hot path code, and how is this affected with the change (i.e. the compiler could already be doing this optimization anyway).

Orthogonal to that, there may be a way to abstract this pattern as: `self.set(x)` i.e. set bit to match the truthiness of x. Not sure how viable is this in Rust.

Reply:
I agree with @pb8o, the former is more readable, and on performance it is very likely in most circumstances to be identical.

Reply:
If we are writing code that deals with bits why not to use bit operations. But yes, it is harder to understand what is going on at a glance. And performance may be not that different. But I still feel that in this particular case less readable but more "suited" variant is better.